### PR TITLE
Dependency validity

### LINKED ISSUE #85: Feature to check dependency cycles
Current implementation plan: create class to perform topological sort.

### PR DESCRIPTION
**Pull request by pengnam**
Resolves #85 
## Features
1) Added functionality to check dependency cycles among tasks.
Refer to `src/main/java/seedu/address/model/DependencyGraph.java` for details on usage.
2) DependencyCommand now checks for dependency cycle before adding dependency, else error is thrown.
3) Topological sort can be performed to display tasks sorted topologically. Function is to be used in list or in aiding the display of appropriate tasks. (@remo5000: for reference)

## New files
src/main/java/seedu/address/model/DependencyGraph.java
src/test/java/seedu/address/model/task/DependencyGraphTest.java

## Files changed
src/main/java/seedu/address/logic/commands/DependencyCommand.java
src/main/java/seedu/address/model/task/exceptions/GraphCycleException.java
src/test/java/seedu/address/logic/commands/DependencyCommandTest.java
src/test/java/seedu/address/testutil/TaskBuilder.java

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/DependencyCommand.java [hunk 1]
```diff
@@ -21,33 +23,43 @@
...
+        //Checking if dependant or dependee already contains dependency
         Task taskDependant = lastShownList.get(dependantIndex.getZeroBased());
         Task taskDependee = lastShownList.get(dependeeIndex.getZeroBased());
         if (taskDependee.getDependency().containsDependency(taskDependee)) {
             throw new CommandException(MESSAGE_ALREADY_DEPENDANT);
         }
-
+        //Checking if introducing dependency will create a cyclic dependency
         Task updatedTask = createDependantTask(taskDependant, taskDependee);
+
```
Comment (unsatisfiedpopcorn) [inline src/main/java/seedu/address/logic/commands/DependencyCommand.java]: Excess blank line

File: src/test/java/seedu/address/logic/commands/DependencyCommandTest.java [hunk 2]
```diff
@@ -49,6 +49,21 @@ public void execute_invalidIndexUnfilteredList_throwsCommandException() {
         assertCommandFailure(dependencyCommand, model, commandHistory, Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
     }
 
+    @Test
+    public void execute_cyclicDependencyUnfilteredList_throwsCommandException() {
+        Task dependantTask = model.getFilteredTaskList().get(INDEX_FIRST_TASK.getZeroBased());
+        Task dependeeTask = model.getFilteredTaskList().get(INDEX_SECOND_TASK.getZeroBased());
+
+
```
Comment (unsatisfiedpopcorn) [inline src/test/java/seedu/address/logic/commands/DependencyCommandTest.java]: excess blank line

File: src/test/java/seedu/address/model/task/DependencyGraphTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        a = DependencyCommand.createDependantTask(a, c);
+        preCyclicTasks.add(a);
+        preCyclicTasks.add(b);
+        cyclicTask = c;
+
+        //CyclicTasks
+        cyclicTasks = new ArrayList<>(preCyclicTasks);
+        cyclicTasks.add(cyclicTask);
+    }
+    @Test
```
Comment (unsatisfiedpopcorn) [inline src/test/java/seedu/address/model/task/DependencyGraphTest.java]: blank line between brace and annotation

**Comment by unsatisfiedpopcorn** [comment]
LGTM, some minor style issues.

**Comment by chikchengyao** [comment]
@pengnam please fix jeff's style comments, then can merge! Algo looks okay to me.

**Pull request merged**